### PR TITLE
session: fix deadlock when init domain failed

### LINKED ISSUE #40408: Dead lock when domain Init failed.
https://github.com/pingcap/tidb/blob/2d755a818c8802cf8fb9f2516ae20c9da3b84fdb/session/tidb.go#L53-L109

We tried to acquire locks in lines L54 and L96(ref L85, onClose func) respectively, resulting in deadlock.

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #40408

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Found by https://github.com/pingcap/tidb/issues/40404 issue, the stack of goroutine is 
```

goroutine 1 [semacquire, 33 minutes]:
sync.runtime_SemacquireMutex(0xc0005aa480?, 0x0?, 0x0?)
	/usr/local/go/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0x7bad560)
	/usr/local/go/src/sync/mutex.go:171 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:90
github.com/pingcap/tidb/session.(*domainMap).Delete(0x7bad560, {0x538e730?, 0xc000467810?})
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/session/tidb.go:112 +0x50
github.com/pingcap/tidb/session.(*domainMap).Get.func1.1()
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/session/tidb.go:85 +0x25
github.com/pingcap/tidb/domain.(*Domain).Close(0xc000068780)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/domain/domain.go:872 +0x18d
github.com/pingcap/tidb/session.(*domainMap).Get.func1()
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/session/tidb.go:96 +0x606
github.com/pingcap/tidb/util.RunWithRetry(0x1e, 0x1f4, 0xc0016509f0)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/util/misc.go:69 +0x89
github.com/pingcap/tidb/session.(*domainMap).Get(0x7bad560, {0x538e730, 0xc000467810})
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/session/tidb.go:76 +0x1d6
github.com/pingcap/tidb/session.createSessionWithOpt({0x538e730, 0xc000467810}, 0x0)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/session/session.go:3122 +0x59
github.com/pingcap/tidb/session.CreateSessionWithOpt({0x538e730, 0xc000467810}, 0xc000c3ec68?)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/session/session.go:2710 +0x3b
github.com/pingcap/tidb/session.CreateSession(...)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/session/session.go:2704
github.com/pingcap/tidb/br/pkg/gluetidb.Glue.UseOneShotSession({{}, {{}}}, {0x538e730, 0xc000467810}, 0x1, 0xc0009f2f90)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/gluetidb/glue.go:119 +0x5e
github.com/pingcap/tidb/br/pkg/task.RunBackup({0x5366770, 0xc000b60c00}, {0x537f180?, 0x7be8c00?}, {0x4c1289f, 0xb}, 0xc00002c700)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/task/backup.go:353 +0x7a8
main.runBackupCommand(0xc000a07b00, {0x4c1289f, 0xb})
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/backup.go:48 +0x418
main.newFullBackupCommand.func1(0xc000a07b00?, {0xc000a16a00?, 0xa?, 0xa?})
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/backup.go:117 +0x25
github.com/spf13/cobra.(*Command).execute(0xc000a07b00, {0xc00021e2a0, 0xa, 0xa})
	/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:916 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc00054ac00)
	/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
main.main()
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/main.go:57 +0x33c
```
After this PR, it will return err1 instead of deadlock.
image
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: session/tidb.go [hunk 1]
```diff
@@ -52,19 +52,21 @@ type domainMap struct {
 
 func (dm *domainMap) Get(store kv.Storage) (d *domain.Domain, err error) {
 	dm.mu.Lock()
```
Comment (hawkingrei) [inline session/tidb.go:54]: use ```syncutil.Mutex to replace ```sync.Mutex``` so that we can detect deadlock in the CI.

**Review by zimulala — APPROVED** [review]
LGTM

File: session/tidb.go (hunk 1 above)
Comment (Defined2014) [reply session/tidb.go:55]: Done, PTAL @hawkingrei

**Review by hawkingrei — APPROVED** [review]
LGTM

**Review by xhebox — APPROVED** [review]

**Comment by hawkingrei** [comment]
/unhold 

some cases just are timeout.

**Pull request merged**